Runtime/Scripts/BaseMeshSync.cs

```diff
@@ -1775,46 +1786,173 @@ EntityRecord UpdateInstancedEntity(TransformData data)
+            {
                 infoRecord = new InstanceInfoRecord();
                 m_clientInstances.Add(path, infoRecord);
             }
-        
+
+            EntityRecord instancedEntityRecord;
             // Look for reference in client instanced objects
-            if (m_clientInstancedEntities.TryGetValue(path, out EntityRecord instancedEntityRecord)) { 
+            if (m_clientInstancedEntities.TryGetValue(path, out instancedEntityRecord))
+            {
                 infoRecord.go = instancedEntityRecord.go;
             }
-            else if (m_clientObjects.TryGetValue(path, out EntityRecord entityRecord)) {
-                infoRecord.go = entityRecord.go;
+            else if (m_clientObjects.TryGetValue(path, out instancedEntityRecord))
+            {
+                infoRecord.go = instancedEntityRecord.go;
             }
-            else {
+            else
+            {
                 throw new Exception($"[MeshSync] No instanced entity found for path {data.path}");
-            } 
-        
-            // Look for parent in client objects
-            if (m_clientObjects.TryGetValue(data.parentPath, out EntityRecord parentRecord)) {
-                if (infoRecord.renderer == null) {
-                    infoRecord.renderer = parentRecord.go.AddComponent<MeshSyncInstanceRenderer>();
-                }
-            
-                infoRecord.renderer.UpdateAll(data.transforms, infoRecord.go);
             }
-            else {
-                if (infoRecord.renderer == null) {
-                    infoRecord.renderer = m_rootObject.gameObject.AddComponent<MeshSyncInstanceRenderer>();
-                }
-            
-                infoRecord.renderer.UpdateAll(data.transforms, infoRecord.go);
+
+            // Find parent for this instance:
+            GameObject instanceRendererParent;
+            if (m_clientObjects.TryGetValue(data.parentPath, out EntityRecord parentRecord))
+            {
+                instanceRendererParent = parentRecord.go;
+            }
+            else
+            {
+                instanceRendererParent = m_rootObject.gameObject;
+
                 Debug.LogWarningFormat(
                     "[MeshSync] No entity found for parent path {0}, using root as parent", data.parentPath);
             }
 
+            switch (InstanceHandling)
+            {
+                case InstanceHandlingType.InstanceRenderer:
+                    infoRecord.DeleteInstanceObjects();
+
+                    if (instanceRendererParent != null && infoRecord.renderer == null)
+                    {
+                        infoRecord.renderer = instanceRendererParent.AddComponent<MeshSyncInstanceRenderer>();
+                    }
+                    infoRecord.renderer.UpdateAll(data.transforms, infoRecord.go);
+                    break;
+                case InstanceHandlingType.Copies:
+                case InstanceHandlingType.Prefabs:
```

Comment:
This switch case is really long. It also seems to be having some checks on the variable, which makes the code hard to follow. Might be better if the code was broken down into smaller functions and the Copies had its own function that reuses the common stuff from the prefabs.

As a first step, it would be great if some of the statement was broken down into smaller functions.

Reply:
Done